Implement no throw version of `unpackTo` to avoid throwing exception on dataplane

tyxia opened the pull request:
Return the error message to the caller site and let caller decides how to handle it (e.g., logging etc) instead of throwing the exception.

Signed-off-by: Tianyu Xia <tyxia@google.com>

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

tyxia commented:
/assign @yanavlasov 

PTAL, Thanks!

yanavlasov reviewed (changes requested):
/wait

yanavlasov commented on source/common/protobuf/utility.cc:
```diff
@@ -562,6 +562,17 @@ void MessageUtil::unpackTo(const ProtobufWkt::Any& any_message, Protobuf::Messag
+absl::Status MessageUtil::unpackToNoThrow(const ProtobufWkt::Any& any_message,
+                                          Protobuf::Message& message) {
+  if (!any_message.UnpackTo(&message)) {
+    return {absl::StatusCode::kInternal, "Unable to unpack as " +
```
```suggestion
    return absl::InternalError(absl::StrCat("Unable to unpack as ", message.GetDescriptor()->full_name(), ": ",
                                            any_message.DebugString());
```

yanavlasov reviewed (changes requested):
/wait

tyxia commented:
/retest

tyxia commented:
/retest

tyxia commented on source/common/protobuf/utility.cc:
```diff
@@ -562,6 +562,16 @@ void MessageUtil::unpackTo(const ProtobufWkt::Any& any_message, Protobuf::Messag
   }
 }
 
+absl::optional<std::string> MessageUtil::unpackToNoThrow(const ProtobufWkt::Any& any_message,
```
Make sense! Done.

tyxia commented:
Thanks for the review. PTAL @yanavlasov 

I have addressed the comments and the only CI failure seems to be caused by test environment (stopped hearing from agent i-0ba525d18868b93b4).

yanavlasov commented on source/common/protobuf/utility.cc:
```diff
@@ -562,6 +562,17 @@ void MessageUtil::unpackTo(const ProtobufWkt::Any& any_message, Protobuf::Messag
+absl::Status MessageUtil::unpackToNoThrow(const ProtobufWkt::Any& any_message,
+                                          Protobuf::Message& message) {
+  if (!any_message.UnpackTo(&message)) {
+    return absl::InternalError(absl::StrCat("Unable to unpack as " +
```
Concatenating strings with + is very inefficient. If replace `+` with `,` the StrCat will make it a lot better.

tyxia replied:
oops, I used strcat but forgot to change the `+`.
Thanks! Done.

yanavlasov reviewed (changes requested):
/wait
